### DISCUSSION TITLE
ContainsAny query optimization

### DISCUSSION THREAD

**Discussion by BernhardGlueck** [ravendb/ravendb#17827] [Q&A]
Hi

I have a question, currently we build a static index where each entry looks like the following

{ 
   NodeIds: ["x","y","z"] 
}

etc.. 

I need to query for all entries with a contains any:

MyNodeIds = [......]

query.Where(q => NodeIds.ContainsAny(MyNodeIds));

So far so good...

However MyNodeIds is stored in another document, and we're basically roundtripping this needlessly... 

e.g Pseudocode:

MyNodeIds = LoadDocument<D1>("..:").MyNodeIds
query.Where(q => q.NodeIds.ContainsAny(MyNodeIds)

MyNodeIds can become quite large ( e.g 4k entries ) 

Is there a way to save the roundtrip e.g.

query.Where(q => q.NodeIds.ContainsAny(LoadDocument<D1>("").MyNodeIds)

so that the node ids stay sever side and need not be transferred ? 
( I know that the above is just pseudo code and that any api for that would look quite different ) 

Also please node that this approach with contains any comes from the fact that 
we have to prevent a cartesian product in our indices, so expanding the index 
to fan out does not help. 
eg...@ayende.com's profile photo

**Comment by ayende** [comment]
That round trip is pretty important, because what you want to requires matching all `MyNodeIds` to all `NodeIds`, which can be very big number.

Can you explain a bit more about your model and what you are trying to do? This is a bit of an X Y Problem

**Reply by BernhardGlueck** [reply]
I think i can explain it with the most similar system I know.. 

Let's assume we have git repository with its trees pointing to content blobs. 
Tree Nodes have a node id and  content. blobs have a content id .. both them are the hashes of their content.

Now we want to search for content, so for that for each unique content blob we would extract some index values ... e.g created at time or file size etc.. 

That is easy with ravendb... we just use an index with dynamic fields. 

Now in addition we would like to filter based on a commit ( or tree root id for that matter ) 

The naive approach would fanout the index for each content based on the commits it is present int .. which for files that are fairly static would give us an index size of O(commit * content

Which is clearly bad.  Since commits do not have linear paths quite often range compressing commit ranges is also not a good strategy.. 

But what we can do is instead of fanning out the index by commit id .. we fan it out by the tree node ids it occurs in.. Since large parts of the repository stay static usually over time, this limits the index growth somewhat since tree nodes are reused quite often..

Now we can build another index that maps commit ids to tree nodes that are referenced by the commit. 

And thus we get my scenario


1. Lookup the commit id, get a list of tree node ids
2. Query the content index for any values and intersect the treenode ids from 1 with the tree node ids in this index.... 

Btw in my example our tree is not an arbitrary tree like in git but a Balanceding B+ Tree with a branching factor of 256.. stored as raven documents.

**Reply by BernhardGlueck** [reply]
Just some numbers... with 256 branching factors for a repository of 1MIO files that gives us 4000 tree node ids to match

**Comment by ayende** [comment]
Classic model is:

```
public class Item
{
   public string ParentId;
}
```

And then you can go up in the tree. That has an issue in a tree model, however, since any update to the parent would result in an update to all the child nodes.

What if you go the other way around?

```
public class Item
{
   public List<string> Children;
}
```

And in the index you go _down_?

The key here is that the children don't change often (won't trigger updates). And instead of a fanout, you can have a simple array of all the matches.

The problem is that a root node will have to go through all of the items, of course. Which may be _very_ big.

Let's talk about something simpler, like a product catalog, we have:

* Categories
* Products

With a many-many association of products and categories. And categories themselves are hierarchical. 

Let's say that we have:

```
{
   "Name": "Ping Pong Ball",
   "Categories": [
       "sports/balls/ping-pong",
       "games/ping-pong"
  ]
}
```

We encode the categories hierarchy in the ids, so we can now ask for anything in the sports involving balls using:

```
`from Items 
where startsWith(Categories, 'sports/balls')
```

Would something like that work for you?
